### ISSUE TITLE
About the output of the posenet

### ISSUE THREAD

**Issue by wangy9711** [anuragranj/cc#7]
hey, I traind all nets accroding to the readme file.When I tested posenet, I found that the camera pose in output was in 6DoF format and then converted to a 3*4 transformation matrix.

In a 3*4 transformation matrix, we usually think that the last column describes the translational motion of the camera.But when the test datas are pictures at the corner of the street,I feel that the network is not performing well when describing the translation.

For example,I chose five images as input .These images are all from the KITTI dataset.They described a street corner. In the transformation matrix of the output, three variables describing the translation, I think at least two variables should have a large change, however, there is only one.  This obviously does not describe the positional change of the camera during the turning of the vehicle. Detailed data are attached.

Where did I go wrong?

Thanks.
[test_data.zip](https://github.com/anuragranj/cc/files/3067860/test_data.zip)

**Comment by anuragranj** [comment]
How did you train it? You can check out the training schedule here: https://github.com/anuragranj/cc/issues/5

**Comment by wangy9711** [comment]
I think I find my mistake.I didn't finish the training process.
Thanks a lot !

**Comment by wangy9711** [comment]
I just traind the net acroding the  training schedule,but i still had some problems.

1. In  TensorBoard->Images ,i can't watch the output image.The output images are all black/red/white.

2. In the training process,errors always fluctuate rather than converge.

Here is my training script:

`python3 train.py $data_root --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 1.0 -pf 0.0 -m 0.0 -c 0.0 -s 0.1 --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 50 --smoothness-type edgeaware --fix-masknet --fix-flownet --log-terminal --name STEP1 --resume

echo "STEP1 done!"

python3 train.py $data_root --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 0.0 -pf 1.0 -m 0.0 -c 0.0 -s 0.1  --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --smoothness-type edgeaware --fix-dispnet --fix-posenet --fix-masknet --log-terminal --name STEP1 --resume

echo "STEP2 done!"

python3 train.py $data_root --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future -b 4 -pc 1.0 -pf 0.5 -m 0.0 -c 0.0 -s 0.1  --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --smoothness-type edgeaware --fix-dispnet --fix-posenet --fix-flownet --log-terminal --name STEP1 --resume

echo "STEP3 done!"



for((i=1;i<=10;i++));  
do


python3 train.py $data_root --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 1.0 -pf 0.5 -m 0.05 -c 0.0 -s 0.1  --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --smoothness-type edgeaware --fix-masknet --fix-flownet --log-terminal --name STEP1 --resume
   
echo $(expr $i );  
echo "LOO1 done!"

python3 train.py $data_root5 --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 0.0 -pf 1.0 -m 0.005 -c 0.0 -s 0.1  --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --smoothness-type edgeaware --fix-dispnet --fix-posenet  --fix-masknet --log-terminal --name STEP1 --resume
echo $(expr $i );  
echo "LOO2 done!"

python3 train.py $data_root --kitti-dir $gt_root --dispnet DispResNet6 --posenet PoseNetB6 --masknet MaskNet6 --flownet Back2Future  -b 4 -pc 1.0 -pf 0.5 -m 0.005 -c 0.3 -s 0.1  --epoch-size 1000 --log-output -f 0 --nlevels 6 --lr 1e-4 -wssim 0.997 --with-flow-gt --with-depth-gt --epochs 100 --smoothness-type edgeaware --fix-dispnet --fix-posenet --fix-flownet --log-terminal --name STEP1 --resume
echo $(expr $i );  
echo "LOO3 done!"

done  
`


btw,I haven't finished the training yet.It takes too much time.But I  want to see the final result.I would appreciate it if you can provide the final well-trained model!

**Comment by anuragranj** [comment]
You should see some reasonable outputs after finishing STEP 1, for the depth and camera motion networks.
The trained models are already provided here:
https://github.com/anuragranj/cc#downloads

**Comment by wangy9711** [comment]
How do you evaluate the global effect?
Can I fuse into a global trajectory based on the pose of every 5 images?
I performed a simple superposition and found that there is a large error in the global trajectory.

**Comment by anuragranj** [comment]
We provide the pose evaluation code for the KITTI odometry dataset. You could use the following as described in the README.
```
python test_pose.py pretrained/pose_model_best.pth.tar --img-width 832 --img-height 256 --dataset-dir /path/to/kitti/odometry/ --sequences 09 --posenet PoseNetB6

```

**Comment by wangy9711** [comment]
Sorry for my ambiguous expressions, I want to get the trajectory in global frame giving a sequence of images. 

I noticed in your works, the output of trajectory is estimated for every 5 images, like 0th-4th, 1st-5th, 2nd-6th. So I wonder how I can get the whole trajectory for all images?

I just do some frame transformation in the original output to transform all images frame to the 0th camera frame, but the result is not really satisfactory. Whatever the sequence is, the trajectory after I processed is just like a straight line.

There is a simple test, used 5 images which is a turning scenes, you can see it in below (It's a right turn scene):
![2](https://user-images.githubusercontent.com/45476795/56465970-df142c00-643b-11e9-91ce-11add0a0b930.png)


And in below pic, the left is groundtruth (The unit is m), right is the output using ```test_pose.py```(the command in your last answer).(I times 100 to the scale of output trajectory).
![1](https://user-images.githubusercontent.com/45476795/56465931-73ca5a00-643b-11e9-822c-6af0e64fa887.png)


It predicts the camera move along an oblique line, and also has a wrong scale. 

I am confused about that, thanks for your patient reply and excellent works!
Waiting for your suggestions.

**Comment by anuragranj** [comment]
The pose network estimates only the relative pose, and not the global pose. The pose network outputs the relative pose between (I--,I), (I-. I), (I+,I) and (I++, I). So, you would have to integrate to get the global pose trajectory. If you just compute the output of the pose network across the sequence of image, you would just get relative pose or the gradient of the global pose trajectory.

**Comment by wangy9711** [comment]
I integrated the output and got the global pose trajectory.
![1](https://user-images.githubusercontent.com/45476795/56850065-b0142380-692f-11e9-9836-682b36f2aa1c.png)
![2](https://user-images.githubusercontent.com/45476795/56850066-b1dde700-692f-11e9-96b4-5770bfdc4831.png)
In the above two images,the red-purple-blue one is groundtruth，the yellow-orange-pink one is integrated trajectory. All data comes from the KITTI data set. It seems that the global performance is not very good....

**Comment by anuragranj** [comment]
Looks like there is an offset. I have a few questions.
1. Did you use the KITTI odometry sequence 9 or sequence 10 to compute this? If no, could you compute it on that?
2. If you used KITTI odometry sequence 9 or 10, could you compute the errors locally and see if it matches the errors that we report which is 
```
 Sequence 9   Sequence 10
0.012 ± 0.007 0.012 ± 0.008
```
These numbers are obtained by running test_pose.py on these sequences.

3. Could you confirm that images are resized to 832x256 are normalized between [-1,1] before they are input to the network?

If 1. 2. 3. is true, then it is possible that our method is not so good on global trajectory although it is good for local relative pose estimation, and a longer temporal information would be required by the network (may be training using LSTMs on the whole sequence) to improve the performance.

**Comment by wangy9711** [comment]
I did above two tests on sequence 7 and 9. But I also did the same test on sequence 10 and got a similar result.
 I also make sure the format of the input is the same as what you described. When I run the script, the output numbers are the same as you report.


The network may have a slight error globally, but it does have excellent results when estimating local pose.Thanks for your patient reply and excellent works again!

**Comment by yyuuliang** [comment]
hi
can u guys share how to output the complete trajectory? I tried several methods, but I couldn't get a smooth result as @wangy9711 
thanks 
![Screenshot from 2020-06-26 23-57-31](https://user-images.githubusercontent.com/21992330/85916791-de79e100-b808-11ea-8b49-fed5f9b8e5dd.png)
![Screenshot from 2020-06-26 23-57-46](https://user-images.githubusercontent.com/21992330/85916793-e174d180-b808-11ea-9624-fa1a5b605840.png)
